### ISSUE TITLE
SearchableRepository

### ISSUE THREAD

**Issue by charlesdaniels** [fyne-io/fyne#2479]
We should evaluate creating a "SearchableRepository" to expose optimized searching functionality via the repository system.

Thanks Martyn!

**Comment by stuartmscott** [comment]
Good idea! Would this use a simple string query, or more complex query interface?

Should also consider whether this would include filters, or if that would be a separate "FilterableRepository". ie. Search "foo", Filter Type "image/jpeg", Filter Modified Since "2021-01-01 12:00", Filter Size Under "10Mb"

**Comment by charlesdaniels** [comment]
This was brought up in the Q&A for FyneConf 2021, so I haven't put much consideration into it.

I think a big difficulty will be the fact that different storage back-ends may support different subsets of those capabilities. Maybe some will have remote-side full-text search and others won't. Some might be able to filter by file size and others may not. Having a unified system for doing queries would be hard if all of those different capabilities are optional. In particular, I think composing multiple queries together is something that is likely to vary by platform.

I find myself wishing that cloud storage providers could support something like [ZFS Channel Programs](https://klarasystems.com/articles/understanding-zfs-channel-programs/?utm_source=rss&utm_medium=rss&utm_campaign=understanding-zfs-channel-programs), where some portion of that API surface would allow you to determine which other portions were available. But we don't live in that timeline.

My instinct would be to have a unified expression language which "compiles" to some kind of intermediate representation describing the query, and then the Fyne runtime would decide based on what the repository has made available how to actually execute the query.

For example, maybe the query is `file_size >= 10*Mb and content_matches /[a-zA-Z0-9]+/`. The repository might support querying by file size, querying by regexing the body text, and may or may not support compound queries. So depending on that, there could be several valid ways for the runtime to realize this query, for example:
* If representable by a compound query, do all the work on the remote
* If not, query by regex match and by file size separately, then do the set intersection on the local end
* If we can query by file size but not regex, do that then pull all the matches down locally to regex through them
* ...

Translating between regex dialects is likely to be another challenge.

Here's a general idea of one way to do this:

```
type SearchableRepository interface{
    Search(QueryFragment) []URI
}

type QueryFragment interface {
    // exactly one of these will return nil, and exactly one a valid pointer
    // what I really want is a tagged union, which I don't know how to do in Go
    Atom() *QueryAtom 
    Scope() *QueryScope
}

type QueryAtom struct{
    Attribute string // somewhere will be a list of allowed attribute, e.g. "file_size"
    Condition operator // an enum or equivalent, like ">", "=", ...
    Match string // whatever the query needs to match based on the operator... how to handle numeric values though?
}

func (a *QueryAtom) Atom() *QueryAtom {
    return a
}

func (a *QueryAtom) Scope() *QueryScope {
    return nil
}

type QueryScope struct{
    Atoms []QueryFragment
    Any bool // select whether everything in the scope has to match, or only some things
    // Maybe instead of Any, we want some more flexible kind of operators?
}

// Atom() and Scope() can be implemented for QueryScope in the obvious way
```

So the Fyne API endpoint would take our query `file_size >= 10*Mb and content_matches /[a-zA-Z0-9]+/` and compile it to something like this:

```
&QueryScope {
    Atoms: []QueryFragment {
        &QueryAtom{Attribute: "file_size", Condition: GreaterEqual, Match: "10000000"},
        &QueryAtom{Attribute: "content", Condition: RegexMatches, Match: "[a-zA-Z0-9]+"},
    },
    Any: false, // combined with an and
}
```

This would then be passed into the repository implementation of `Search()`, which could process this representation and try to execute it in the most optimal way.

I'm just spitballing here, but maybe it's a start.
